proposals/pattern-variables.md

```diff
@@ -0,0 +1,73 @@
+# Variable declarations under disjunctive patterns
+
+## Summary  
+
+Allow variable declarations under `or` patterns and across `case` labels in a `switch` section.
+
+## Motivation
+
+This feature would reduce code duplication where we could use the same piece of code if either of patterns is satisfied. For instance:
+```cs
+if (e is (int x, 0) or (0, int x))
+    Use(x);
+  
+switch (e)
+{
+    case (int x, 0):
+    case (0, int x):
+        Use(x);
+        break;
+}
+```
+Instead of:
+
+```cs
+if (e is (int x1, 0))
+    Use(x1);
+else if (e is (0, int x2))
+    Use(x2);
+  
+switch (e)
+{
+    case (int x, 0):
+        Use(x);
+        break;
+    case (0, int x):
+        Use(x);
+        break;
+}
+```
+
+## Detailed design
+
+Variables *must* be redeclared under all disjuncitve patterns because assignment of such variables depend on the order of evaluation which is undefined in the context of pattern-matching.
+
+- In a *disjunctive_pattern*, pattern variables declared under the left-hand-side must be redeclared under the right-hand-side.
```

Comment:
The way this is worded, technically I could declare variables under just the right-hand side. Perhaps "variables declared on one side must be redeclared on the other side"?